File: packages/gateway/kotlin/gateway/src/main/kotlin/com/bridgething/gateway/LocalLogRelay.kt
package com.bridgething.gateway

public object LocalLogRelay {
    @Volatile private var sink: ((String, String, String) -> Unit)? = null

    public fun setSink(sink: ((String, String, String) -> Unit)?) {
        this.sink = sink
    }

    /**
     * Whether anything is listening. Producers check this before doing work to
     * build a line: with no stream open nothing may cross the native boundary,
     * and parsing every logcat line for a sink that will drop it is pure drain.
     */
    public fun hasSink(): Boolean = sink != null

    public fun push(level: String, target: String, message: String) {
        sink?.invoke(level, target, message)
    }
}

File: packages/gateway/swift/Sources/BridgethingGateway/LocalLogRelay.swift
import Foundation

public final class LocalLogRelay: @unchecked Sendable {
  public static let shared = LocalLogRelay()

  private let lock = NSLock()
  private var sink: (@Sendable (String, String, String) -> Void)?

  private init() {}

  public func setSink(_ sink: (@Sendable (String, String, String) -> Void)?) {
    lock.lock()
    defer { lock.unlock() }
    self.sink = sink
  }

  public func push(level: String, target: String, message: String) {
    lock.lock()
    let sink = self.sink
    lock.unlock()
    sink?(level, target, message)
  }
}
